geps/gep-1731/index.md

```diff
@@ -0,0 +1,436 @@
+    // MUST not exceed the Request timeout duration. If any retry attempts are
+    // still in progress when the Request timeout duration has been reached,
+    // these SHOULD be cancelled if possible and the Gateway MUST immediately
+    // return a timeout error.
+    //
+    // If a BackendRequest timeout (`rules[].timeouts.backendRequest`) is
+    // configured on the route, any retry attempts which reach the configured
+    // BackendRequest timeout duration without a response SHOULD be cancelled if
+    // possible and the Gateway should wait for at least the specified backoff
+    // duration before attempting to retry the backend request again.
+    //
+    // If a BackendRequest timeout is _not_ configured on the route, retry
+    // attempts MAY time out after an implementation default duration, or MAY
+    // remain pending until a configured Request timeout or implementation
+    // default duration for total request time is reached.
+    //
+    // When this field is unspecified, the time to wait between retry attempts
+    // is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Backoff *Duration `json:"backoff,omitempty"`
+}
+
+// HTTPRouteRetryStatusCode defines an HTTP response status code for
+// which a backend request should be retried.
+//
+// Implementations MUST support the following status codes as retryable:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// Implementations SHOULD support the `"5xx"` shorthand for matching status
+// codes in the 500-599 range. For implementations supporting `"5xx"`, this
+// shorthand MUST match the following status codes:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// The `"5xx"` shorthand MAY additionally match a well-documented subset of
+// arbitrary status codes in the 500-599 range, or match all status codes in
+// this range.
+//
+// Implementations SHOULD NOT support retrying status codes in the 100-399
+// range, as these responses are generally not appropriate to retry.
+//
+// Implementations MAY support specifying discrete values in the 400-499 range,
+// which are often inadvisable to retry, and MAY support discrete values in the
+// 600-999 (inclusive) range, which are not valid for HTTP clients, but are
+// sometimes used for communicating application-specific errors.
+//
+// +kubebuilder:validation:Pattern=`^[1-9](?:[0-9][0-9]|xx)$`
+type HTTPRouteRetryStatusCode string
+
+// Duration is a string value representing a duration in time. The foramat is
+// as specified in GEP-2257, a strict subset of the syntax parsed by Golang
+// time.ParseDuration.
+//
+// +kubebuilder:validation:Pattern=`^([0-9]{1,5}(h|m|s|ms)){1,4}$`
+type Duration string
+```
+
+### YAML
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: HTTPRoute
+metadata:
+  name: retry-example
+spec:
+  ...
+  rules:
+  - backendRefs:
+    - name: some-service
+      port: 8080
```

Comment:
```suggestion
  rules:
  - matches:
    - path:
        value: "/foo"
      headers:
      - name: "version"
        value: "v2"
    - path:
        value: "/v2/foo"
    backendRefs:
    - name: foo-v2
      port: 8080
      weight: 90
    - name: foo-v2-canary
      port: 8080
      weight: 10
```
Expanding the example to be a more complex HTTPRouteRule with multiple matches and backends to clarify the `retry` stanza will apply to all matches and backends within a given rule.

/cc @linsun